### ISSUE TITLE
Log Grouping on GitHub Actions

### ISSUE THREAD

**Issue by chrissimon-au** [slaveOftime/Fun.Build#55]
Github Actions supports [grouping log entries](https://docs.github.com/en/actions/using-workflows/workflow-commands-for-github-actions#grouping-log-lines) by emitting "workflow commands".

It would be quite nice if you could opt in to have the workflow commands emitted at the start and end of top-level stages (or pipelines).  (They don't yet support [nested groupings](https://github.com/actions/toolkit/issues/1001) but once that's done, grouping pipelines _and_ stages would be good.)

A few ideas on what this could look like...

### Option to log before and after each stage

```
pipeline "name" {
   beforeEachStageEcho (fun ctx -> $"::group::{ctx.Name}")
   afterEachStageEcho (fun ctx -> $"::endgroup::")
}
```

### Option to run arbitrary code before and after each stage

More flexible, but maybe more than is needed?

```
pipeline "name" {
   beforeEachStage (fun ctx -> ctx.RunCommand($"echo ::group::{ctx.Name}")
   afterEachStage (fun ctx -> ctx.RunCommand($"echo ::endgroup::")
}
```

### Log group optimised

More restricted, but would still allow to take advantage of other log grouping, e.g. [Azure Devops Piplines](https://learn.microsoft.com/en-us/azure/devops/pipelines/scripts/logging-commands?view=azure-devops&tabs=bash#formatting-commands) has a similar feature.

```
pipeline "name" {
   logGroups "::group::{name}" "::endgroup::"
}
```

### Github Actions specific

Far more restrictive, but easiest to configure!

```
pipeline "name" {
   githubActionsLogGroups true
}
```

**Comment by chrissimon-au** [comment]
I did experiment with an approach for doing this by composing nested stages:

```
let logGroup (s: Internal.StageContext) =
    stage s.Name {
        echo $"::group::{s.Name}"
        s
        echo "::endgroup::"
    }
```

Used like so:

```
pipeline "name" {

   logGroup <| stage "Name" {
      echo "here"
   }
}
```

But it wasn't ideal - when collapsed it looks like:

![image](https://github.com/slaveOftime/Fun.Build/assets/11002331/7af9ec3c-e613-480c-a805-7e4e5b93302c)

When not collapsed, it looks like:

![image](https://github.com/slaveOftime/Fun.Build/assets/11002331/159aa78b-5ce5-4ca7-afc6-b8c9ea60efc9)

The extra nested stages are redundant, and there is a lot of stage info logged before and after the collapse, which clutters things.

**Comment by albertwoo** [comment]
@chrissimon-au nice idea. And I like the "Option to run arbitrary code before and after each stage", implemented like below:

```fsharp
pipeline "test" {
    description "Format code and run tests"
    runBeforeEachStage (fun ctx -> if ctx.GetStageLevel() = 0 then printfn $"::group::{ctx.Name}")
    runAfterEachStage (fun ctx -> if ctx.GetStageLevel() = 0 then printfn "::endgroup::")
    stage_checkEnv
    stage_lint
    stage_test
    runIfOnlySpecified
}
```

It should be available in 1.0.5. 🎉

**Comment by chrissimon-au** [comment]
Nice, thanks for the quick turnaround!
